DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,146 @@
+    int npart = particle_positions.size() / ndim;
+    particle_positions = std::move(pp);
+    particle_velocities = std::move(pv);
+    best_particle_positions = std::vector<double>(npart * ndim);
+    best_position = std::vector<double>(ndim);
+};
+
+
+void ParticleSwarmState::addParticlesInsideBox(const std::vector<double> &box_lower, const std::vector<double> &box_upper,
+                                               const std::function<double(void)> get_random01) {
+    checkVarSize("box lower bounds", box_lower.size(), num_dimensions);
+    checkVarSize("box upper bounds", box_upper.size(), num_dimensions);
+    for (int i=0; i<num_particles; i++) {
+        for (int j=0; j<num_dimensions; j++) {
+            double range = std::fabs(box_upper[j] - box_lower[j]);
+            int idx = i * num_dimensions + j;
+            particle_positions[idx] = range * get_random01() + box_lower[j];
+            particle_velocities[idx] = 2 * range * get_random01() - range;
+        }
+    }
+    initialized = true;
+}
+
+
+void ParticleSwarm(ObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
+
+    // Only run the algorithm on properly initialized states.
+    if (!state.initialized) {
+        throw std::runtime_error("Particles have not been initialized in the input state object");
+    }
+
+    // Initialize helper variables and functions.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> &particle_positions = state.getParticlePositionsRef();
+    std::vector<double> &particle_velocities = state.getParticleVelocitiesRef();
+    std::vector<double> &best_particle_positions = state.getBestParticlePositionsRef();
+    std::vector<double> &best_position = state.getBestPositionRef();
+    std::vector<double> prev_best_vals(num_particles), cur_best_vals(num_particles);
+    std::vector<bool> prev_is_inside(num_particles), cur_is_inside(num_particles);
+    double best_fval = std::numeric_limits<double>::max();
+    ObjectiveFunctionConstrained f_constrained = makeObjectiveFunctionConstrained(num_dimensions, f, inside);
+
+    // Create a lambda that updates the best particle and swarm positions based on function values and domain information.
+    auto update_best_positions = [&](const std::vector<double> prev_vals, const std::vector<double> cur_vals,
+                                     const std::vector<bool> prev_domain, const std::vector<bool> cur_domain)->void {
+        for (size_t i=0; i<num_particles; i++) {
+            bool is_smaller = (cur_vals[i] < prev_vals[i]) || !prev_domain[i];
+            if (cur_domain[i] && is_smaller) {
+                std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
```

Comment:
`std::copy_n` is a cleaner way to do this, you don't need the middle iterator, e.g.,:
```
std::copy_n(particle_positions.begin() + i * num_dimensions, num_dimensions, best_particle_positions.begin() + i * num_dimensions);
```